File: 316-remove-duplicate-letters/316-remove-duplicate-letters.cpp
class Solution {
public:
    string removeDuplicateLetters(string s) {
        
        int n = s.length();
        vector<bool> visit(26, false);
        vector<int> lastIndex(26, 0);
        
        // 1. set last indexes for every char
        for(int i=0; i<n; i++){
            lastIndex[s[i]-'a'] = i;
        }
        
        stack<char> st;
        
        // 2. iterate over string s
        for(int i=0; i<n; i++){
            
            if(visit[s[i]-'a']) continue;
            
            visit[s[i] - 'a'] = true;
            
            while(st.size()>0 && (st.top() > s[i]) && lastIndex[st.top()-'a'] > i){
                char ch = st.top();
                st.pop();
                visit[ch-'a'] = false;
            }
            
            st.push(s[i]);
        }
        
        string res="";
        while(!st.empty()){
            res += st.top();
            st.pop();
        }
        
        reverse(res.begin(), res.end());
        return res;
    }
};

File: 316-remove-duplicate-letters/316-remove-duplicate-letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        
        n = len(s)
        visit = set()
        lastIndex = [0]*26
        
        for i in range(n):
            lastIndex[ord(s[i])-ord("a")] = i
        
        st = []
        for i in range(n):
            
            if s[i] in visit:
                continue
            
            visit.add(s[i])
            
            while len(st)>0 and ord(st[-1]) > ord(s[i]) and lastIndex[ord(st[-1]) -ord("a")]>i:
                visit.remove(st.pop())
            
            st.append(s[i])
            
        res = ""
        while st:
            res += st[-1]
            st.pop()
            
        res = res[::-1]
        return res
    